Feature/order : Issue #28, #29

seongkyu-lim opened the pull request:
- 주문 생성 로직 리팩토링 
- 추가된 엔티티들과의 관계 매핑
- 리뷰 받은 것들에 대한 코드 리팩토링을 진행하였습니다.

자세한 구현사항은 이슈에 목록으로 정리해 놨습니다 :)

😅조회 API들에 대하여 해당 엔티티의 값만 조회하고 연관된 엔티티는 가져오지 않도록 우선 설계가 되어있는데, 의도한 것이 아니며, 임의로 해놨음을 알려드립니다. 추후 요구사항 및 필요여부에 따라 조회 시 얻을 수 있는 정보를 수정하겠습니다..!

😭수정된 로직에 대해서도 api testing tool을 진행하였습니다, 테스트 결과는 이슈 #29에 올려놨습니다.

seongkyu-lim committed:
add : getOrderTableList API

요구사항 : 주문 테이블의 목록을 조회할 수 있다.

seongkyu-lim committed:
add : enum ( OrderType, OrderStatus )

seongkyu-lim committed:
refactor: mapping deliveryAddress and order entity.

order(many) to deliveryAddress(one)

seongkyu-lim committed:
rebase

seongkyu-lim committed:
refactor: dto

seongkyu-lim committed:
refactor: issue #28 final commit.

- #28 이슈관련 수정 및 마무리.

seongkyu-lim committed:
refactor: logic about order, orderTable, orderMenu

- 주문 생성 로직 변경.
- 그에 따른 엔티티 값 수정.
- 주문 테이블, 주문 엔티티 일대일 매핑.
등등..!

seongkyu-lim committed:
refactor: ktlint

humanitas03 commented on src/main/kotlin/com/kitchenforce/domain/delivery/DeliveryAddressRepository.kt:
```diff
@@ -1,7 +1,11 @@
-interface DeliveryAddressRepository : JpaRepository<DeliveryAddress, Int>
+interface DeliveryAddressRepository: JpaRepository<DeliveryAddress, Int> {
+
+    fun findByAddress(address : String): DeliveryAddress?
```
주소(address)가 unique한 속성일까요?
조회 결과가 2개 이상 나오게 되면 에러가 날 것 같습니다.

humanitas03 commented on src/main/kotlin/com/kitchenforce/domain/enum/OrderStatus.kt:
```diff
@@ -0,0 +1,12 @@
+package com.kitchenforce.domain.enum
+
+enum class OrderStatus(
+    val id: Int,
+    val title: String,
+    val description: String
+) {
+    WAITING(0, "주문 대기", "주문 대기 상태"),
+    ACCEPTED(1, "주문 접수", "주문 접수 상태"),
+    CLOSE(2, "주문 종료", "주문 종료 상태"),
+    ERROR(3, "에러", "")
```
enum에 id, title, description 까지 명시한 특별한 이유가 있을까요?
WATING("주문대기") 정보 정도만 있어도 충분하지 않을까 생각이 듭니다.

humanitas03 commented on src/main/kotlin/com/kitchenforce/service/OrderService.kt:
```diff
@@ -0,0 +1,124 @@
+
+    fun getOrder(userId: Long): OrderDto {
+
+        val orderDtoNothing: OrderDto = OrderDto(
+            userId = userId,
+            orderType = OrderType.ERROR,
+            orderStatus = OrderStatus.ERROR,
+            paymentMethod = "",
+            requirement = "",
+            deliveryAddress = "",
+            orderTableDto = null,
+            orderMenuDtoList = ArrayList()
+        )
```
조회 결과가 없을 때 내려주는 Dto를 셋업 해주고 있는데, 결과가 없으면 Client에 ERROR로 내려주는게 적당한지 생각해볼 필요가 있을 것 같습니다.

클라이언트 입장에서 DB에 없는 주문을 요청해서 나온 결과인지, 서버의 문제로 인해 정상적으로 조회가 되어야 하는데 결과가 ERROR로 내려오는지 헷깔릴수 있는 문제일 것 같습니다.

seongkyu-lim replied:
네, NotFoundException으로 처리하여 정확한 문구를 클라이언트에게 알려주는 것이 적합한 것 같습니다. 감사합니다.😊

humanitas03 commented on src/main/kotlin/com/kitchenforce/domain/enum/OrderType.kt:
```diff
@@ -0,0 +1,12 @@
+    DELIVERY(0, "배달", "배달 주문"),
+    TAKEOUT(1, "포장", "포장 주문"),
+    EATIN(2, "매장", "매장 취식 주문"),
+    ERROR(3, "에러", "")
```
OrderStatus에서도 동일하게, 에러(ERROR)라고 하는 값을 OrderType Enum으로 관리 해줄 필요가 있을까요?

humanitas03 commented on src/main/kotlin/com/kitchenforce/service/OrderService.kt:
```diff
@@ -0,0 +1,124 @@
+import org.springframework.transaction.annotation.Transactional
+
+@Service
+class OrderService(
```
다소 추상적인 내용일수 있습니다.

객체지향 SOLID 원칙 중에 단일 책임 원칙(SRP, Single Responsibility Priciple) 이라는 원칙이 있는데요, 
어떤 클래스가 가진 책임이 여러개가 되면 변화에 취약해지고 유연성이 떨어지므로 가급적 하나의 클래스는 하나의 책임만을 담당하는 것이 바람직하다 라는 것이 해당 원칙의 요지입니다.

제가 볼 때 ```OrderService```가 다소 많은 책임을 가지고 있는것으로 보입니다. 주문 이라고 하는 업무가 크게 테이블 주문과 배달 주문으로 구분이 될 것으로 보입니다.  두 주문이 처리되는 프로세스도 차이가 있고, 이를  ```create(dto: OrderDto)``` 라고 하는 하나의 함수에서 처리하려고 하면 함수 내부에서 각각의 주문의 속성에 따라서 DB에 대한 CRUD를 분기를 쳐야 될 것같습니다.

기능은 유지하되 좀 더 나은 설계를 고려한다면 세부적인 주문을 처리하는 Service 클래스를 하위에 더 나누고,(DeliveryOrderService, EatInOrderService 등등...) 지금의 create 메서드는 파사드(Facade)로 두는 방식을 사용하는게 어떤가 싶네요.

seongkyu-lim replied:
좋은 말씀 감사합니다, 지연님이 맡아서 해주시는 배달과 관련해 배달의 상태를 처리하는 로직도 필요해서 결국 배달 주문 서비스도 필요하니, 말씀 주신대로 주문 서비스를 배달, 테이블 주문으로 나누어 주문 생성 로직도 세분화 하는 것이 좋은 것 같습니다.👍🏻

humanitas03 commented on src/main/kotlin/com/kitchenforce/service/OrderService.kt:
```diff
@@ -0,0 +1,124 @@
+                    )
+                }
+            )
+        } ?: return orderDtoNothing
```
이것도 결과가 없을때 ```ProductException``` 에서 처리했던것 처럼 예외로 던져서 ExceptionHandler를 태우는 방식으로 처리해보는것 어떨까요?

humanitas03 reviewed (commented):
로컬에서 실 REST API 호출로 테스트를 진행해보신것 같은데요,
REST는 JUnit의 MockMvc 테스트를, Controller 이외의 기능들에 대해서 JUnit 단위 테스트 코드 작성을 권해 드립니다.

humanitas03 commented on src/main/kotlin/com/kitchenforce/service/OrderService.kt:
```diff
@@ -0,0 +1,124 @@
+            orderTableRepository.save(orderTable)
+        }
+
+        val orderMenuList: MutableList<OrderMenu> = ArrayList()
+        for (orderMenuDto in dto.orderMenuDtoList) {
+            val orderMenu: OrderMenu = OrderMenu(
+                quantity = orderMenuDto.quantity,
+                order = order,
+                menu = menuRepository.findByName(orderMenuDto.menuName)
+            )
+            orderMenuList.add(orderMenu)
+        }
+        orderMenuRepository.saveAll(orderMenuList)
+    }
```
이전에도 잠깐 언급을 드렸었는데요, Java의 stream() 메서드를 사용하는 것에 익숙하다면, 
List같은 컬렉션을 for문보다는 함수형으로 로직을 사용해보는 것도 한번 검토해보시는걸 권해드립니다.

45 ~ 55 까지의 코드 라인을 아래로 대체 가능할 것 같습니다.
```kotlin
        dto.orderMenuDtoList.map{ OrderMenu(
            quantity = it.quantity,
            order = order,
            menu = menuRepository.findByName(it.menuName)
        )}.also{
            orderMenuRepository.saveAll(it)
        }

```

seongkyu-lim replied:
also 함수를 이렇게도 사용할 수 있군요, 감사합니다!

seongkyu-lim committed:
refactor: service logic, enum

-  반복문 함수형 로직으로 리팩토링.
- null인 경우 빈 값을 반환하는 것이 아닌 exception으로 처리.
- enum클래스 error 상태 삭제.

seongkyu-lim committed:
Merge branch 'develop' into feature/#28

neverlish commented on src/main/kotlin/com/kitchenforce/controller/OrderController.kt:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    @GetMapping("/{userId}")
+    fun getOrder(@PathVariable userId: Long): OrderDto {
```
/ID 에 들어가는 인자는, 일반적으로는 레코드 ID 가 들어가는 경우가 많더라고요.(getById의 느낌)

특정 사용자 ID의 주문들을 가져오는 기능이라면, URL을 따로 만들어보는건 어떨가요?

seongkyu-lim committed:
refactor: enum

seongkyu-lim committed:
delete: userId

seongkyu-lim committed:
Merge remote-tracking branch 'origin/feature/#28' into feature/#28

# Conflicts:
#	src/main/kotlin/com/kitchenforce/domain/delivery/DeliveryAddressRepository.kt
